csrc/opaque_type.h

```diff
@@ -0,0 +1,71 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <any>
+#include <functional>
+#include <ostream>
+
+namespace nvfuser {
+
+class Opaque;
+
+template <typename T>
+struct OpaqueEquals {
+  bool operator()(const Opaque& a, const Opaque& b) const;
+};
+
+class Opaque {
+  std::any value_;
+  std::function<bool(const Opaque&, const Opaque&)> equals_;
+
+ public:
+  template <typename T>
+  explicit Opaque(T value)
+      : value_(std::move(value)), equals_(OpaqueEquals<T>{}) {}
+
+  bool operator==(const Opaque& other) const {
+    if (this == &other) {
+      return true;
+    }
+    if (value_.type() != other.value_.type()) {
+      return false;
+    }
+    return equals_(*this, other);
+  }
```

Comment:
Right now, we manage `equals_` by ourselves. Users are not allowed to change or customize it. So the same `T` now always guarantee to have the same `equals_`. There is no longer a concept of `lhs`'s `equals_` or `rhs_`'s `equals_`, there is only a `T`'s `equals_`.

So as long as the underlying type `T` has symmetric `==`, it is impossible to have asymmetry here. On the other hand, even if `T` does has an asymmetric operator `==`, then: 1. this is a bad programming habit, and it is the user's problem, not `Opaque`'s problem. 2. even though 1, what the user want is still respected, because we are calling `equals_(*this, other)` instead of `equals_(other, *this)`

Reply:
Makes sense. Thanks.